### DISCUSSION TITLE
ListBoxItem - WhenActivated

### DISCUSSION THREAD

**Discussion by ctrl-alt-d** [AvaloniaUI/Avalonia#6427] [Q&A] [answered]
In a `ListBox` only the firsts `ListBoxItem` raise `WhenActivate`. Looks related to virtualization. How can I raise `WhenActivate` for all items?

```c#

public class CentreRowUserCtrl : ReactiveUserControl<CentreRowViewModel>
{

  public CentreRowUserCtrl()
  {
    this.WhenActivated(d => { // Only firsts items activated, IDK why.
        d(ViewModel!.ShowDialog.RegisterHandler(UpdateShowDialogAsync));
        System.Console.WriteLine( $"Activating {ViewModel!.Etiqueta}"  );
    });
    AvaloniaXamlLoader.Load(this);            
  }
```

In this screenshot you can see how firsts `ListBoxItem` are activateds (because they are a `System.Console.WriteLine` ) and they appear on console, but when I scroll down, no more items are activated:

![activating](https://user-images.githubusercontent.com/3105983/129547982-4c66b651-ae82-44bc-90ac-42c6fee63243.gif)

Is there a way to activate all items?

I have not this issue if I use an `ItemsControl` instead of a `ListBox`, but, with an `ItemsControl` performance is unacceptable.

Thanks!

**Comment by ctrl-alt-d** [comment]
As a workaround, I replaced ReactiveUI by this:

```c#
    public class CentreRowUserCtrl : ReactiveUserControl<CentreRowViewModel>
    {
        public CentreRowUserCtrl()
        {
            AvaloniaXamlLoader.Load(this);
            // this.WhenActivated(d => d(ViewModel!.ShowDialog.RegisterHandler(UpdateShowDialogAsync)));
        }

        protected override void OnDataContextEndUpdate()
        {
            base.OnDataContextBeginUpdate();
            ViewModel?.ShowDialog.RegisterHandler(UpdateShowDialogAsync);
        }
```

**Comment by kekekeks** [comment]
It seems that WhenActivated doesn't handle DataContext updates and is only triggered by AttachedToVisualTree event. So if control changes its data context (e. g. when it's reused by a virtualized list) the event is not fired.

cc: @worldbeater

**Reply by ctrl-alt-d** [reply]
Thanks @kekekeks about your comment, finally the solution is to [observe ViewModel it self](https://github.com/AvaloniaUI/Avalonia/discussions/6427#discussioncomment-1191558).

**Comment by maxkatz6** [comment]
Not related to your specific question, but:

> I have not this issue if I use an ItemsControl instead of a ListBox, but, with an ItemsControl performance is unacceptable.

Please, research how these controls work internally before comparing their performance. 
Both ListBox and ItemsControl use ItemsPresenter internally. The main difference ListBox is a selection control, while ItemsControl is not. Also ListBox has virtualization enabled by default, which improves performance and has side effects like this one, where "attached to the visual tree" is raised once per item container.

**Reply by maxkatz6** [reply]
You technically can have the same performance with ItemsControl or using ItemsPresenter directly. But you will miss selection functionality.

**Reply by ctrl-alt-d** [reply]
I just tried `ListBox` vs `ItemsPresenter` performance both with `VirtualizationMode="Simple"` and the same 200 items observable list.

The result is `ListBox` is faster. I don't know why. I can do more tests if you want. Thanks @maxkatz6 , I didn't know about `ItemsPresenter` it's not in Avalonia Controls Docs. (I know about `ItemsControl`, but  `ItemsControl` doesn't have `VirtualizationMode` property)


 Thanks again @maxkatz6 about your comments. I'm learning. All comments are welcome!!!

**Comment by worldbeater** [comment, marked as answer]
As stated in [`WhenActivated` documentation](https://www.reactiveui.net/docs/handbook/when-activated/), the `WhenActivated` ReactiveUI feature is a way to track disposables. There is no guarantee that the `ViewModel` property (the `DataContext` property of an `IViewFor`) won't be `null` when the control is activated, and that the view model won't change since control activation. That's why the `IViewFor.ViewModel` property is marked as nullable in ReactiveUI. The general recommendation is to never access the `ViewModel` property directly inside a `WhenActivated` block. Always use a call to `WhenAnyValue` in a similar fashion to this code snippet:
```cs
this.WhenActivated(disposables => {
    this.WhenAnyValue(x => x.ViewModel)
        .Subscribe(vm => { /* Here you can access the view model safely. */ })
        .DisposeWith(disposables);
});
```
Another option is to use ReactiveUI bindings:
- [`BindInteraction`](https://www.reactiveui.net/docs/handbook/interactions/binding-interactions) for binding types like `Interaction<TInput, TOutput>`,
- [`BindCommand`](https://www.reactiveui.net/docs/handbook/data-binding/#types-of-bindings) for binding `ICommand` implementations,
- [`Bind`](https://www.reactiveui.net/docs/handbook/data-binding/#types-of-bindings) and [`OneWayBind`](https://www.reactiveui.net/docs/handbook/data-binding/#types-of-bindings) for binding properties.

Additionally, consider taking a look at the [Avalonia.NameGenerator](https://github.com/avaloniaui/avalonia.NameGenerator/) tool that simplifies the process of creating code-behind bindings by generating type-safe references to named XAML controls.

**Reply by ctrl-alt-d** [reply]
Hi @worldbeater ,

thanks about your detailed answer, awesome, it works like a charm.

This is my first MVVM app also my first Avalonia App. I'm following Avalonia Docs and Avalonia Music Store and ToDo sample, also reading ReactiveUI docs. If they are other learning resources, please, let me know. I really appreciate your help. Best regards.
